drivers/sensor/iis328dq/iis328dq.c

```diff
@@ -0,0 +1,460 @@
+#ifdef CONFIG_IIS328DQ_THRESHOLD
+	case SENSOR_ATTR_LOWER_THRESH:
+	case SENSOR_ATTR_UPPER_THRESH:
+		if (chan != SENSOR_CHAN_ACCEL_XYZ) {
+			LOG_ERR("Threshold cannot be set per-channel");
+			return -ENOTSUP;
+		}
+		return iis328dq_set_threshold(dev, attr == SENSOR_ATTR_LOWER_THRESH, val);
+	case IIS328DQ_ATTR_DURATION:
+		if (chan != SENSOR_CHAN_ACCEL_XYZ) {
+			LOG_ERR("Duration cannot be set per-channel");
+			return -ENOTSUP;
+		}
+		return iis328dq_set_duration(dev, val->val1);
+#endif /* CONFIG_IIS328DQ_THRESHOLD */
+	default:
+		LOG_DBG("Acc attribute not supported");
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int iis328dq_attr_set(const struct device *dev, enum sensor_channel chan,
+			     enum sensor_attribute attr, const struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ALL:
+		return iis328dq_dev_config(dev, chan, attr, val);
+	default:
+		LOG_DBG("Attr not supported on %d channel", chan);
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int iis328dq_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct iis328dq_data *iis328dq = dev->data;
+	const struct iis328dq_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int16_t buf[3];
+
+	/* fetch raw data sample */
+	if (iis328dq_acceleration_raw_get(ctx, buf) < 0) {
+		LOG_DBG("Failed to fetch raw data sample");
+		return -EIO;
+	}
+
+	iis328dq->acc[0] = buf[0] >> 4;
+	iis328dq->acc[1] = buf[1] >> 4;
+	iis328dq->acc[2] = buf[2] >> 4;
+
+	return 0;
+}
+
+static const struct sensor_driver_api iis328dq_driver_api = {
+	.attr_set = iis328dq_attr_set,
+#if CONFIG_IIS328DQ_TRIGGER
+	.trigger_set = iis328dq_trigger_set,
+#endif /* CONFIG_IIS328DQ_TRIGGER */
+	.sample_fetch = iis328dq_sample_fetch,
+	.channel_get = iis328dq_channel_get,
+};
+
+static int iis328dq_init(const struct device *dev)
+{
+	struct iis328dq_data *iis328dq = dev->data;
+	const struct iis328dq_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t wai;
+
+	iis328dq->dev = dev;
+
+	/* check chip ID */
+	if (iis328dq_device_id_get(ctx, &wai) < 0) {
+		return -EIO;
+	}
+
+	if (wai != IIS328DQ_ID) {
+		LOG_ERR("Invalid chip ID");
+		return -EINVAL;
+	}
+
+	/* reset device */
+	if (iis328dq_boot_set(ctx, PROPERTY_ENABLE) < 0) {
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(100));
+
+	if (iis328dq_boot_get(ctx, &wai) < 0) {
+		return -EIO;
+	}
+	if (wai != PROPERTY_DISABLE) {
+		LOG_ERR("BOOT did not deassert");
+		return -EIO;
+	}
+
+	if (iis328dq_block_data_update_set(ctx, PROPERTY_ENABLE) < 0) {
+		return -EIO;
+	}
+
+	/* set default odr to 12.5Hz acc */
+	if (iis328dq_set_odr(dev, 12) < 0) {
+		LOG_ERR("odr init error (12.5 Hz)");
+		return -EIO;
+	}
+
+	if (iis328dq_set_range(dev, cfg->range) < 0) {
+		LOG_ERR("range init error %d", cfg->range);
+		return -EIO;
+	}
+
+#ifdef CONFIG_IIS328DQ_TRIGGER
```

Comment:
This is wrong. If you have two instances of the same driver you may still want to enable interrupt on one and not on the other. Please take a look to lis2du12 driver and the usage of trig_enabled.

Reply:
I've added a check for `!cfg->gpio_int1.port && !cfg->gpio_int2.port` in `iis328dq_init_interrupt()`, that should have the same effect without introducing a redundant field to the config struct.

Reply:
Please leave to the reviewer the task of marking something as "resolved", otherwise it's a mess.

Anyway, I see your point and seems a correct approach. Only, I believe you need to add same check in trigger_set() routine as well.

Reply:
`trigger_set()` will eventually fail with `-ENOTSUP` in that case anyway, but I've added a fail-fast path.

Reply:
> `trigger_set()` will eventually fail with `-ENOTSUP` in that case anyway, but I've added a fail-fast path.

I think I'm missing why it sould fail anyway, but glad that you add the check.